block-producer/src/rpc/mod.rs

```diff
@@ -0,0 +1,118 @@
+use std::{future::Future, sync::Arc};
+
+use async_trait::async_trait;
+use tokio::sync::{
+    mpsc::{error::SendError, unbounded_channel, UnboundedReceiver, UnboundedSender},
+    oneshot,
+};
+
+/// Creates a client/server pair that communicate locally using channels
+pub fn create_client_server_pair<Request, Response, S>(
+    server_impl: S
+) -> (RpcClient<Request, Response>, RpcServer<Request, Response, S>)
+where
+    Request: Send + 'static,
+    Response: Send + 'static,
+    S: ServerImpl<Request, Response>,
+{
+    let (sender, recv) = unbounded_channel::<(Request, oneshot::Sender<Response>)>();
+
+    let client = RpcClient {
+        send_requests: sender,
+    };
+
+    let server = RpcServer {
+        recv_requests: recv,
+        server_impl: Arc::new(server_impl),
+    };
+
+    (client, server)
+}
+
+/// Errors related to the RPC mechanism itself
+/// TODO: Make errors more descriptive
+#[derive(Debug)]
+pub enum RpcError {
+    SendError,
+    RecvError,
+}
+
+impl<T> From<SendError<T>> for RpcError {
+    fn from(_send_error: SendError<T>) -> Self {
+        Self::SendError
+    }
+}
+
+/// Implements the processing of an RPC request on the server.
+/// Every request is processed in a new task.
+#[async_trait]
+pub trait ServerImpl<Request, Response>: Send + Sync + 'static {
+    async fn handle_request(
+        self: Arc<Self>,
+        x: Request,
+    ) -> Response;
+}
+
+// RPC SERVER
+// --------------------------------------------------------------------------------------
+
+pub struct RpcServer<Request, Response, S>
+where
+    Request: Send + 'static,
+    Response: Send + 'static,
+    S: ServerImpl<Request, Response>,
+{
+    recv_requests: UnboundedReceiver<(Request, oneshot::Sender<Response>)>,
+    server_impl: Arc<S>,
+}
```

Comment:
It is necessary, because this is how all tasks within the txpool/block producer will communicate. For any message that needs to be sent between 2 tasks, an `RpcClient`/`RpcServer` pair will be used. And three things can vary:
1. the request type
2. the response type
3. how the server turns the request into a response

Hence the 3 generic types